### DISCUSSION TITLE
Mitmproxy and TLS1.1

### DISCUSSION THREAD

**Discussion by NeoProgramming** [mitmproxy/mitmproxy#6185] [Q&A]
It seems to me that mitmproxy does not work correctly with TLS1.1

While working with mitmroxy (the end goal is to change the User-Agent for all outgoing https requests), I encountered a situation where some requests through mitmproxy failed.
I did the analysis and wrote a synthetic test: Go program that downloads a small file from the Internet and at the same time directly controlled the TLS version of HTTPS request.

So, if this program downloads the file directly, then it works with all TLS versions: 1.1, 1.2, 1.3
But if I download the file through mitmproxy, then it is downloaded only with TLS versions: 1.2 and 1.3
With version 1.1, a "TLS handshake failed" error is thrown

From which I conclude that the server can work with all versions of TLS, and the problem is with mitmproxy.
I don't care what version the request from mitmproxy to the server will be made with; but in a real situation, apparently, some program accesses through mitmproxy and uses TLS1.1, which leads to an error.

What would you advise to do in this situation? OS is Windows 10. 
The end goal, as I said, is to replace all User-Agents in all requests. This is necessary for Android Studio and Gradle to work on a computer in a local network with certain Internet restrictions (all packages that do not contain the word "chrome" in the User-Agent are blocked :) ).

**Comment by mhils** [comment]
Did you try setting [`tls_version_client_min`](https://docs.mitmproxy.org/stable/concepts-options/#tls_version_client_min)?

**Comment by NeoProgramming** [comment]
Yes of course. For TLS_2, TLS_3 a correct error message is issued:
`Client TLS handshake failed. Client and mitmproxy cannot agree on a TLS version to use. You may need to adjust mitmproxy's tls_version_client_min option.`
But for SSL3, TLS1, TLS1_1 the following error is thrown:
`Client TLS handshake failed. The client may not trust the proxy's certificate for dl.google.com (OpenSSL Error([('SSL routines', '', 'no suitable signature algorithm')]))`
( I am trying to download file https://dl.google.com/dl/android/maven2/androidx/drawerlayout/drawerlayout/1.0.0/drawerlayout-1.0.0.aar )

**Reply by mhils** [reply]
There are two options here, I don't know which one is correct:

1. You can maybe adjust `ciphers_client` to enable TLS 1.1 ciphersuites and then things work?
2. We're shipping with a too-recent OpenSSL that does not support TLS 1.1 anymore?

I'd hope the first option is correct, but I honestly don't know. 😄

**Comment by felixq** [comment]
I confirm that Mitmproxy does not work with tls versions lower than 1.2.
I was diagnosing an old program in .net 4 and couldn't get it to connect to Mitmproxy, .net 4 by default does not work with tls 1.2.
I started trying all versions of tls and ssl lower than 1.2 with curl and it didn't work either.
I tried with ciphers_client, tls_version_client_min and tls_version_client_max.

**Comment by jpstotz** [comment]
Most likely this is caused by the used OpenSSL version. Since OpenSSL 3.1 old TLS modes are disabled by default:

> SSL 3, TLS 1.0, TLS 1.1, and DTLS 1.0 only work at security level 0.

https://github.com/openssl/openssl/blob/master/NEWS.md#major-changes-between-openssl-30-and-openssl-310-14-mar-2023

Not sure how to set security level to 0 in python for the libraries used by mitmproxy.
